common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationFactory.java

```diff
@@ -0,0 +1,171 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.spark.sql.catalyst.util;
+
+import java.util.*;
+import java.util.function.BiFunction;
+import java.util.function.Function;
+
+import com.ibm.icu.util.ULocale;
+import com.ibm.icu.text.Collator;
+
+import org.apache.spark.SparkIllegalArgumentException;
+import org.apache.spark.unsafe.types.UTF8String;
+
+/**
+ * Static entry point for collation aware string functions.
+ * Provides functionality to the UTF8String object which respects defined collation settings.
+ */
+public final class CollationFactory {
+  /**
+   * Entry encapsulating all information about a collation.
+   */
+  public static class CollationInfo {
+    public final String collationName;
+    public final Collator collator;
+    public final Comparator<UTF8String> comparator;
+
+    /**
+     * Version of the collation. This is the version of the ICU library used to create the collator.
+     * For non-ICU collations (e.g. UTF8 Binary) the version is set to "1.0".
+     */
+    public final String version;
+
+    /**
+     * Collation sensitive hash function. Output for two UTF8Strings will be the same if they are
+     * equal according to the collation.
+     */
+    public final Function<UTF8String, Integer> hashFunction;
```

Comment:
should this hash function return long? And for performance, we can use `ToLongFunction[UTF8String]` to avoid java boixing.

Reply:
Sure. At this point `UTF8String` provides 32bit hash, but yeah, we can extend this to `long`. Nice trick with `ToLongFunction`.